Add Command-graph class section (#4)
reble/SYCL-Docs: adoc/extensions/sycl_khr_command_graph.adoc

----
khr::node_type::host_task
----

> a@ Represents a SYCL <<host-task>> command.

Author: Is that not implicit from the fact that we're talking about the type of a node in a command-graph in this table? Maybe I'm not really seeing what more information such a change would provide.

> a@ Represents a SYCL <<host-task>> command.

Reviewer: So I was thinking that host task command could refer to a command that has been submitted, so clarifying that this is a command within a graph, so yet to be submitted, but maybe being a bit too lawyery, and this would be inferred.

> a@ Represents a SYCL <<host-task>> command.

Author: I can change it to align more with the other table entries, something like "Represents a SYCL <<host-task>> execution", would that be better?

|====

[[sec:khr-graph-command-graph]]
== Command-graph class

The [code]#command_graph# class represents a directed acyclic graph of nodes,

> where each node represents a single command with zero or more dependencies

Reviewer 2: It could also be an empty node or subgraph, in that case the node doesn't represent a command or represents multiple commands.
```suggestion
where each node represents zero or more commands with zero or more dependencies
```

Author: @reble Do you have any thoughts on this since you raised this question? Given the discussion I would still be in favour of the "single command" or "single operation" wording here.

Reviewer 2: Consistency is my main concern here. In the node section we define an empty node as:

> An empty node contains no commands.

If a command can consist of zero or more operations, we should update the description.

Author: A command is defined as part of the base SYCL spec already so I don't know that we can augment its definition in that way. I'll just go with the originally suggested wording.

Reviewer 2: Yes I was not questioning that part, but our empty node description needs to be updated.
Can do that as a follow up.

between them.
For the list of supported node types see
<<subsec:khr-graph-supported-node-types>>.

A [code]#command_graph# is associated with a single SYCL device and context on
which nodes representing device commands will be executed.
The execution of a graph is considered complete when all of its nodes have
completed.

> A [code]#command_graph# is built up by either recording queue submissions or

Author: Replaced with "one or both of" 👍

> explicitly adding nodes, then once the user is happy that the graph is complete,

Reviewer 3: "user" bit doesn't really add extra value, could make this more concise  "then once ~~the user is happy that~~ the graph is complete

the graph instance is finalized into an executable variant - a fixed snapshot of
the graph which can no longer have nodes added to it.
Finalization may be a computationally expensive operation as the runtime may be
able to perform optimizations based on the graph structure.
However it can also perform work ahead of time that would normally be needed
immediately prior to command execution.
After finalization the graph can be submitted for execution on a queue one or

> more times with reduced launch overhead.

Author: This might be part of a broader conversation but this is an example of something I think we do a lot in the vendor extension which is describing usage in API definitions. I think the SYCL spec is a lot more segregated. For example in the architecture section they first describe something like the memory model with buffers and accessors, then the programming interface section for buffers is much more focused on the actual class and C++ usage.

To me this type of general description of what the extension does and how to use it belongs more in a general usage/description than in the class definition. Either this is just repeating what we've already said previously, or this is information hidden away in the API definitions that should be more easily available.

This is the default template argument for the [code]#command_graph# class.

a@
[source]
----
graph_state::executable
----

a@ A [code]#command_graph# in this state is a fixed snapshot of the state and
topology of the modifiable graph it was finalized from. Subsequent changes to
that modifiable graph have no effect on the executable graph(s) created from
it.
A graph in this state is ready for execution, but can no longer have new nodes
or edges added to it.
|====

=== Command-graph class interface

[source,,linenums]
----
include::{header_dir}/extensions/khr_graph/command_graph.h[lines=4..-1]
----

[[subsec:khr-graph-command-graph-ctors]]
=== Constructors

All [code]#command_graph# constructors take a parameter named [code]#propList#
which allows the application to pass zero or more properties.
These properties may specify additional effects of the constructor and resulting
[code]#command_graph# object.
See <<subsec:khr-graph-command-graph-properties>> for the [code]#command_graph#
properties that are defined by this extension.

.[apititle]#Context/device constructor#
[source, role=synopsis,id=api:khr-command-graph-ctx-dev-ctor]
----

> command_graph(const context& syclContext, const device& syclDevice,
>                 const property_list& propList = {})

Reviewer 3: ```suggestion
command_graph(const context& syclContext, const device& syclDevice,
              const property_list& propList = {})
```
Nipicking that it would be nice for the params to line up in rendered view. Right now
![image](https://github.com/user-attachments/assets/82047368-5ebb-48b8-a03b-bcc0603cc9dc)

syclQueue.get_device(), propList)#

[[subsec:khr-graph-command-graph-member-funcs]]
=== Member functions

.[apidef]#khr::command_graph::finalize#
[source, role=synopsis,id=api:khr-command-graph-finalize]
----
command_graph<graph_state::executable> finalize(
      const property_list& propList = {}) const
----
_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.

_Effects:_ Synchronous operation that creates a new [code]#command_graph# in the
executable state with a fixed topology that can be submitted for execution on
any queue sharing the device and context associated with the graph.
It is valid to call this method multiple times to create subsequent executable
graphs.
It is also valid to continue to add new nodes to the modifiable graph instance
after calling this function.
It is valid to finalize an empty graph instance with no recorded commands.
Zero or more properties can be provided to this function via an instance of
[code]#property_list#.
Valid properties for this function are defined in
<<subsec:khr-graph-command-graph-properties>>.

_Returns:_ A new [code]#command_graph# object in the executable state.

> _Throws:_ An [code]#exception# with the [code]#errc::invalid# error code if the

Author: I think if a new SYCL command type gets added that we don't support (that doesn't map to an existing command type anyway) that should get caught when you try to use it since we'll fail to map it to a node type. But yeah if we support the command but the specific used backend doesn't we won't know until finalize.

We could possibly add a UR query for supported command types that we could query on graph construction to check against, then we should be able to throw errors always on add. Though if there are any UR command entry points that only have partial support for some features of it we wouldn't know about that until finalize unless that granularity is encoded in the UR query, which could get messy.

> _Throws:_ An [code]#exception# with the [code]#errc::invalid# error code if the

Reviewer: I see what you mean, since record & replay is state of the queue, additional commands or backend support could be introduced which would implicitly be unsupported by record & replay, unless the extensions which introduce those commands are compatible with the SYCL graph extension. This is a tricky relationship to define in formal wording.

I'd prefer that for the KHR we require that a (complete) implementation support all commands supported by the SYCL 2020, would this be possible or are there some limitations to this in the specification currently? If we could do this then an implementation is still free to throw a an exception with the feature not implemented error, but that's more of an implementation detail, and the spec doesn't really need to say anything about it.

Though for other commands which are introduced by orthogonal extensions, we can't expect any new vendor or KHR extension to be compatible with record & replay (at least while SYCL graphs is a KHR and not core SYCL), so it's fair that in this case it will throw an exception, in this case I'd say it makes sense to be an invalid error.

Maybe this is a situation where it would help to have wording which adds to existing APIs; for example we can add wording which describes that the queue can now be in a recording state, and then define what happens for the queue member functions when it's in that state, but we could also say that if any queue member functions do not specify their behaviour when in the recording state, and so calling them in that state results in throwing an exception.

I think it makes sense if possible to throw on calling the queue member functions, rather than waiting to finalise, as it makes it clearer to users where the error came from.

graph contains a command that is not supported by the device.

.[apititle]#khr::command_graph::begin_recording#
[source, role=synopsis,id=api:khr-command-graph-begin-recording]
----

> void begin_recording(queue& recordingQueue,                     (1)
>                        const property_list& propList = {})
> 
> void begin_recording(const std::vector<queue>& recordingQueues, (2)
>                        const property_list& propList = {})

Reviewer 3: ```suggestion
void begin_recording(queue& recordingQueue,                     (1)
                     const property_list& propList = {})

void begin_recording(const std::vector<queue>& recordingQueues, (2)
                     const property_list& propList = {})
```

----
_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.

> _Effects (1):_ Synchronously changes the state of [code]#recordingQueue# to the

Reviewer: I would combine these into a single clause which refers to each overload separately.

void end_recording(const std::vector<queue>& recordingQueues) (3)
----
_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.

_Effects (1):_ Synchronously finishes recording on all queues that are recording
to the graph and sets their state to [code]#queue_state::executing#.
This operation is a no-op for any queue in the graph that is already in the
[code]#queue_state::executing# state.

_Effects (2):_ Synchronously changes the state of [code]#recordingQueue# to the
[code]#queue_state::executing# state.
This operation is a no-op if [code]#recordingQueue# is already in the
[code]#queue_state::executing# state.

_Effects (3):_ Synchronously changes the state of each queue in
[code]#recordingQueues# to the [code]#queue_state::executing# state.
This operation is a no-op for any queue in [code]#recordingQueues# that is
already in the [code]#queue_state::executing# state.

_Throws (2)(3):_

* An [code]#exception# with the [code]#errc::invalid# error code if any queue is
recording to a different graph.

.[apititle]#khr::command_graph::add (empty node)#
[source, role=synopsis,id=api:khr-command-graph-add-empty]
----

> node add(const property_list& propList = {})

Reviewer: For the enqueue commands KHR (https://github.com/KhronosGroup/SYCL-Docs/pull/644) it's being proposed internally that we move to a handler-less interface, effectively removing the need for command groups. If we do go this route and this is adopted in the KHR it may be suggested that for SYCL graphs we also move away from using handlers.

I don't think we need to do anything immediately for this but something to be aware of.

Author: Good to know! Yeah I think we can evaluate that when the design is finalized. I think John (?) said he was going to present it to the WG soon (this week maybe?).

----
_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.

_Effects:_ This creates an empty node which contains no command.
Empty nodes can be used to create transitive dependencies between other graph
nodes, similar to a barrier.
Zero or more properties can be provided to this function via an instance of
[code]#property_list#.
Valid properties for this function are defined in
<<subsec:khr-graph-node-properties>>.

_Returns:_ The empty node which has been added to the graph.

_Throws:_ An [code]#exception# with the [code]#errc::invalid# error code if a
queue is recording commands to the graph.

.[apititle]#khr::command_graph::add#
[source, role=synopsis,id=api:khr-command-graph-add]
----

> template <typename T>
>   node add(T cgf, const property_list& propList = {})

Reviewer 3: ```suggestion
template <typename T>
node add(T cgf, const property_list& propList = {})
```

----
_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.

> _Effects:_ The [code]#cgf# command group function behaves in much the same way

Reviewer 3: Not sure if "Much" adds anything other than ambiguity
```suggestion
_Effects:_ The [code]#cgf# command group function behaves in the same way
```

as the command group function passed to [code]#queue::submit# unless explicitly
stated otherwise.
Code in the function is evaluated synchronously, before the function returns
back to [code]#command_graph::add#, with the exception of any SYCL commands
(e.g. kernels, or explicit memory copy operations).
These commands are captured into the graph and executed asynchronously when the
graph is submitted to a queue.
The requisites of [code]#cgf# will be used to identify any dependent nodes in
the graph to form edges with.
Zero or more properties can be provided to this function via an instance of
[code]#property_list#.
Valid properties for this function are defined in
<<subsec:khr-graph-node-properties>>.

_Returns:_ The node which has been added to the graph.

_Throws:_

* An [code]#exception# with the [code]#errc::invalid# error code if a queue is
  recording commands to the graph.
* An [code]#exception# with the [code]#errc::invalid# error code if the graph
  wasn't created with the
  [code]#property::command_graph::assume_buffer_outlives_graph# property and

>   this command uses a buffer.

Reviewer: I'd maybe rephrase this as the command using an accessor rather than a buffer, as the buffer isn't used directly within the command group.

 
.[apititle]#khr::command_graph::make_edge#
[source, role=synopsis,id=api:khr-command-graph-make-edge]
----
void make_edge(node& src, node& dest)
----
_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.

_Effects:_ Creates a dependency between two nodes representing a happens-before
relationship.
Node [code]#dest# will be dependent on [code]#src#.

_Throws:_

* An [code]#exception# with the [code]#errc::invalid# error code if a queue is
  recording commands to the graph object.

* An [code]#exception# with the [code]#errc::invalid# error code if [code]#src#
  or [code]#dest# are not valid nodes assigned to the graph object.

* An [code]#exception# with the [code]#errc::invalid# error code if [code]#src#
  and [code]#dest# are the same node.

* An [code]#exception# with the [code]#errc::invalid# error code if the
  resulting dependency would lead to a cycle in the graph.
  This error is omitted when [code]#property::command_graph::no_cycle_check# is
  set.

.[apititle]#khr::command_graph::print_graph#
[source, role=synopsis,id=api:khr-command-graph-print-graph]
----

> void print_graph(std::string path, bool verbose = false) const

Reviewer: Having SYCL interface with file system paths could be complicated, maybe we should consider instead having `print_graph` instead print to a std::string, so then the user can decide what to do with it.

Reviewer 3: That's a fair point, but to use the tools that work on the dot language you probably want a file, and right now this feature is very useful. It's really easy to add to your application for debugging without having to write it to a file yourself. 

We could have 2 versions of this API one that write to a std::string if the current API throws an exception because it can't write to a file?

----
_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.

> _Effects:_ Synchronous operation that writes a DOT formatted description of the

Reviewer 3: reminder to say whether we overwrite if a file already exists on the path

Author: I think our current behaviour is to overwrite, but is that actually what we want? We could codify it either way I think, I don't know what's better. We could error by default and have a bool param to force overwrite if the file already exists?

I guess we could also add like a UTC timestamp or equivalent to the end of the filename to avoid overwrites, that might be good too.

Reviewer 3: I would agree that I don't think it matters too much, all these suggestions are viable.

If your printing a graph inside multiple application steps then I think a user does something like this to get the iteration they want:
```
static unsigned iterations = 0;
if (iterations == myIteration) {
graph.print("/tmp/mygraph.dot");
}
iterations++;
```

And if they want to see all the iterations, then they'll need to suffix "iterations" to the file path to get dump of each iteration, where the iteration number would probably be more useful than a timestamp in the filename. Although a small amount more work on their part to do  `std::string("/tmp/mygraph_") + std::to_string(iterations) + std::string(".dot")`

bool parameter with default sounds good to me for controling whether to overwrite or not.
